### PR TITLE
Implement support for defining new symbols in CORE.e.setting

### PR DESCRIPTION
**Pull request by vrurg**
Because loading of all CORE settings caused some tests to fail due to the way `SETTING` and `CORE` pseudo-packages behaved, this fix also includes improvements to `PseudoStash` class. Most notable changes are:

- `PSEUDO::<missing-symbol>` now returns `Failure` where previously it was `Nil`. Done so to conform to the way `::()` works.
- `LEXICAL::` now includes all symbols from scope's lexical chains and dynamic symbols from its caller chain. This conforms to the rule "all symbols visible at location"
- `SETTING` now includes all symbols from all `CORE.setting`s.
- `CORE` now includes few additional symbols installed by `World.loading_and_symbol_setup`. Not sure about them, but seem to be harmless and doesn't affect roast though was breaking `rakudo` test.
- `UNIT` now includes all lexicals visible at the unit level.

- Also binding via chained pseudos is now supported:

  ```
  my $*foo;
  sub do-bind { CALLERS::<$*foo> := 42 }
  do-bind;
  say $*foo; # 42
  ```

- `EVAL` doesn't load `CORE.setting` anymore. Effectively, `use v6.X` inside `EVAL`s is just ingnored. If this approach is not getting re-considered I would suggest making it a `X::Language::TooLate` exception.

   Also `EVAL` context now gets !EVAL_MARK` symbol alongside with `!UNIT_MARK`. It is necessary to allow `SETTING` pseudo find the real unit.

- `die` and `fail` now work with `$!` as default. The symbol is been searched in `CALLER::LEXICAL::` as `S29-context/die.t` spectest suggests.

- new sub in `CORE`: `CORE-SETTING-REV`. Returns currently active language revision. It's purpose is to support rakudo tests.

- Fixed a hidden bug where `request` was always installing submodules into its `CALLER::MY::` context. The bug was hidden until pseudo-packages started returning `Failure` for missing symbols.

### CODE REVIEW

**Comment by vrurg** [comment]
Fix for #3028

File: src/Perl6/Actions.nqp [hunk 1]
```diff
@@ -6613,35 +6615,26 @@ class Perl6::Actions is HLL::Actions does STDActions {
...
+            my $numsemis := +$<semilist><statement>;
             if $numsemis > 1 {
                 $past := QAST::Stmts.new( :node($/) );
                 $past.push(QAST::Op.new( :op('call'), :name('&infix:<,>'), :node($/)));
             }
-
-            my $semi := -1;
-            while ++$semi < $numsemis {
-                my $EXPR := @EXPR[$semi];
+            my $semi := 0;
```
Comment (lizmat) [inline src/Perl6/Actions.nqp]: Feels like you missed a commit (a26e95bab41e7b787c) that changed this.  Is there a reason you want to go back to the old behaviour here?

File: src/Perl6/Actions.nqp [hunk 2]
```diff
@@ -6882,6 +6863,59 @@ class Perl6::Actions is HLL::Actions does STDActions {
...
+        # Look for any chained adverb pairs and relocate them.
+        # Try to reuse existing QAST where possible.
+        elsif $*FAKE_INFIX_FOUND {
+            my $numsemis := +$<semilist><statement>;
+            if $numsemis > 1 {
+                $past := QAST::Stmts.new( :node($/) );
+                $past.push(QAST::Op.new( :op('call'), :name('&infix:<,>'), :node($/)));
+            }
+            my $semi := 0;
+            repeat until $semi >= $numsemis {
```
Comment (lizmat) [inline src/Perl6/Actions.nqp]: Same comment here, why revert a26e95bab41e7b787c8 ?

File: src/Perl6/Optimizer.nqp [hunk 3]
```diff
@@ -20,7 +20,7 @@ my class Symbols {
     # Some interesting scopes.
     has $!GLOBALish;
     has $!UNIT;
-    has $!SETTING;
+    # has $!SETTING;
```
Comment (lizmat) [inline src/Perl6/Optimizer.nqp]: Why comment out when it is not needed anymore?

File: src/Perl6/Optimizer.nqp [hunk 4]
```diff
@@ -260,28 +260,47 @@ my class Symbols {
...
 
+    # method find_in_setting($symbol) {
+    #     if !nqp::defined($!SETTING) {
+    #         my int $i := +@!block_stack;
+    #         while $i > 0 && !nqp::defined($!SETTING) {
+    #             $i := $i - 1;
+    #             my $block := @!block_stack[$i];
+    #             my %sym := $block.symbol("!CORE_MARKER");
+    #             if +%sym {
+    #                 $!SETTING := $block;
```
Comment (lizmat) [inline src/Perl6/Optimizer.nqp]: NQP does support `last`.  Put it here, then you could remove the `!defined($!SETTING)` from the while condition for more simple code.

File: src/Perl6/Optimizer.nqp [hunk 5]
```diff
@@ -260,28 +260,47 @@ my class Symbols {
...
-                }
-            }
-            if !nqp::defined($!SETTING) {
-                nqp::die("Optimizer couldn't find CORE while looking for $symbol.");
-            }
-        } else {
-            if nqp::existskey(%!SETTING_CACHE, $symbol) {
-                return %!SETTING_CACHE{$symbol};
-            }
+        if nqp::existskey(%!SETTING_CACHE, $symbol) {
```
Comment (lizmat) [inline src/Perl6/Optimizer.nqp]: Am I seeing the same method twice ??

File: src/Perl6/Actions.nqp (hunk 1 above)
Comment (vrurg) [reply src/Perl6/Actions.nqp]: That's confusing. The commit is actually there vrurg/rakudo@a26e95bab41e7b787c8283f7ae24ae3859727cfe. Perhaps a result of chained migration macbook -> mac mini -> macbook with help of backups and rsync.

File: src/Perl6/Optimizer.nqp (hunk 3 above)
Comment (vrurg) [reply src/Perl6/Optimizer.nqp]: Temporary state of things. Explanation is below.

File: src/Perl6/Optimizer.nqp (hunk 5 above)
Comment (vrurg) [reply src/Perl6/Optimizer.nqp]: The previous version is commented out. Mine is still active. I wasn't sure if my version is not flawed and temporarily left the old one in the comments.

Thanks, Liz!

BTW, this all is coming from attempt to make `QuantHash`es iterables in 6.e.

File: src/Perl6/Actions.nqp (hunk 1 above)
Comment (vrurg) [reply src/Perl6/Actions.nqp]: I have re-applied related commits by cherry-picking them.

**Comment by jnthn** [comment]
> MY:: now includes all symbols from its lexical chains and dynamic symbols from its caller chain

Hm, I'd expect `MY::` to only include lexical things (with the proviso that `my $*FOO` is actually a lexical)?

**Comment by jnthn** [comment]
> SETTING is now includes all symbols from all CORE.settings. Not sure if this is right. 

Good question. I guess if it doesn't, then we'd want a `SETTINGS` that did. What does `CORE` do with regard to this question? Include everything from all the CORE.setting.*?

File: src/core/PseudoStash.pm6 [hunk 6]
```diff
@@ -25,10 +27,8 @@ my class PseudoStash is Map {
     my $pseudoers := nqp::hash(
         'MY', sub ($cur) {
             my $stash := nqp::clone($cur);
-            nqp::bindattr_i($stash, PseudoStash, '$!mode', PRECISE_SCOPE);
-            nqp::setwho(
-                Metamodel::ModuleHOW.new_type(:name('MY')),
-                $stash);
+            nqp::bindattr_i($stash, PseudoStash, '$!mode', STATIC_CHAIN +| DYNAMIC_CHAIN);
```
Comment (jnthn) [inline src/core/PseudoStash.pm6]: I think `PRECISE_SCOPE` is right for `MY`.

File: src/core/PseudoStash.pm6 [hunk 7]
```diff
@@ -74,17 +74,13 @@ my class PseudoStash is Map {
                 nqp::bindattr($stash, Map, '$!storage', nqp::ctxlexpad($ctx));
                 nqp::bindattr($stash, PseudoStash, '$!ctx', $ctx);
                 nqp::bindattr_i($stash, PseudoStash, '$!mode', PRECISE_SCOPE);
```
Comment (jnthn) [inline src/core/PseudoStash.pm6]: Since I think really `MY::` is meant to just be like `OUTER::`, but without walking a step of the chain.

**Comment by vrurg** [comment]
> > MY:: now includes all symbols from its lexical chains and dynamic symbols from its caller chain
> 
> Hm, I'd expect `MY::` to only include lexical things (with the proviso that `my $*FOO` is actually a lexical in terms of its storage)?

That'd be my guess too. But according to `S02-names/pseudo.t`, lines 100+, `MY` is expected to include dynamics from its outer. In a way it makes sense if we interpret `MY` as what is visible in the current scope.

Making `MY` PRECISE_SCOPE breaks the test. So, we either shall remove it from the roast or leave things as they're in this PR.

**Comment by vrurg** [comment]
> > SETTING is now includes all symbols from all CORE.settings. Not sure if this is right.
> 
> Good question. I guess if it doesn't, then we'd want a `SETTINGS` that did. What does `CORE` do with regard to this question? Include everything from all the CORE.setting.*?

This is my weak point here as many aspects are not clear to me yet. `CORE`, I suppose, must do exactly what you say. But what would make it different from `SETTINGS` then?

**Comment by vrurg** [comment]
> Hm, I'd expect `MY::` to only include lexical things (with the proviso that `my $*FOO` is actually a lexical in terms of its storage)?

Another uncertainty:

```
sub foo {
    my $foo;

    {
        say MY::<$foo>;
    }
}
```

Shall this fail? It will if `MY` is `PRECISE_SCOPE`.

**Comment by vrurg** [comment]
The top comment has been updated to reflect latest changes suggested by @jnthn and @lizmat

**Comment by vrurg** [comment]
Related topic: perl6/problem-solving#58

**Comment by vrurg** [comment]
At this point any all known bugs revealed by this PR were fixed, all spectests are passing with perl6/roast#557 applied. I consider this PR ready for merge.

Ping @jnthn

**Pull request merged**